h2o-core/src/main/java/water/rapids/Rapids.java

```diff
@@ -92,121 +94,293 @@ public static Val exec(String rapids, Session session) {
     }
   }
 
+
+  //--------------------------------------------------------------------------------------------------------------------
+  // Private
+  //--------------------------------------------------------------------------------------------------------------------
+
+  // Set of characters that cannot appear inside a token
+  private static HashSet<Character> invalidTokenCharacters =
+          new HashSet<>(Chars.asList("({[]}) \t\r\n\\\"\'".toCharArray()));
+
+  // Set of characters that may appear in a number. Note that "NaN" or "nan" is also a number.
+  private static HashSet<Character> validNumberCharacters =
+          new HashSet<>(Chars.asList("0123456789.-+eEnNaA".toCharArray()));
+
+  /**
+   * The constructor is private: rapids expression can be parsed into an AST tree, or executed, but the "naked" Rapids
+   * object has no external purpose.
+   * @param rapidsStr String containing a Rapids expression.
+   */
+  private Rapids(String rapidsStr) {
+    _str = rapidsStr;
+    _x = 0;
+  }
+
   /**
-   * Parse an expression
+   * Parse and return the next expression from the rapids string.
    * '('   a nested function application expression ')
    * '{'   a nested function definition  expression '}'
-   * '#'   a double: attached_token
    * '['   a numeric list expression, till ']'
    * '"'   a String (double quote): attached_token
    * "'"   a String (single quote): attached_token
    * digits: a double
    * letters or other specials: an ID
    */
-  public AstRoot parse() {
+  private AstRoot parseNext() {
     switch (skipWS()) {
-      case '(':  return new AstExec(this); // function application
-      case '{':  return new AstFunction(this);  // function definition
-      case '#':  _x++;                     // Skip before double, FALL THRU
```

Comment:
What about Flow code (see `h2o-flow` repo)? 

I was tracking old Rapids discussions:
```
Question:
     #number         // leading '#' then a standard Java Double <-- why have an alternate form to bare numbers?

Answer:
No reason except to ease backwards compatiblity.  Don't have to change all of the R wrapper all in one go.  But can totally drop this alternative form.  Requires literally 1 line of code to allow this in the backend.
```

Seems ok to drop.